### DISCUSSION TITLE
How to define custom polymorphic QuerySets and Managers?

### DISCUSSION THREAD

**Discussion by fed239** [jazzband/django-polymorphic#696] [Q&A] [answered]
I'd like to add `filter_by_user` methods to all QuerySet classes defined for child polymorphic models and use them to filter objects of every model when called for the base model. `filter_by_user` methods may be different for different child models. Is this possible?

I tried like in the following code but it didn't work.

```
class MyBaseQuerySet(PolymorphicQuerySet):
    def filter_by_user(self, user):
        return self.all()


class MyBaseModel(PolymorphicModel):
    ...
    objects = PolymorphicManager(MyBaseQuerySet)


class MyChild1QuerySet(MyBaseQuerySet):
    def filter_by_user(self, user):
        return self.filter(fieldA__lt=user.profile.propA)


class MyChild1Model(MyBaseModel):
    fieldA = models.IntegerField()
    ...
    objects = PolymorphicManager(MyChild1QuerySet)


class MyChild2QuerySet(MyBaseQuerySet):
    def filter_by_user(self, user):
        return self.filter(fieldB__gt=user.profile.propB)


class MyChild2Model(MyBaseModel):
    fieldB = models.IntegerField()
    ...
    objects = PolymorphicManager(MyChild2QuerySet)
```

When I call `MyBaseModel.objects.filter_by_user(john)` magic doesn't happen and only base queryset `filter_by_user` is called.

**Comment by dsummersl** [comment]
I think I'm seeing the same thing for this scenario:

    class MyBaseModel(PolymorphicModel):
        ...
        # no "objects =" defined.

    class SubclassFeedManager(PolymorphicManager):
        def get_queryset(self):
            ...

    class MySubModel(MyBaseModel):
        objects = SubclassFeedManager()


    ...

    MySubModel.objects.all()

This call doesn't use the SubclassFeedManager() at all.

**Comment by MastaChimp** [comment]
Can somebody please suggest a workaround for this issue? I also want the custom manager to be called on each of the child classes when you call the base class custom manager.

**Comment by bckohan** [comment, marked as answer]
Hello @fed239 thanks for your question - sorry its been a decade for someone to get back to you!

There are a few concepts to understand here. First - the difference between a manager and a queryset. ```Model.objects``` is a Manager instance. Methods on that manager instance may return ```QuerySet``` instances (e.g. all). Which brings me to the first point:

* If you write a custom ```QuerySet``` - the methods you add to that ```QuerySet``` will not be available on the manager by default unless you create a manager class from the QuerySet.

There are several ways to specify custom managers and querysets:
1. [PolymorphicManager.from_queryset()](https://docs.djangoproject.com/en/stable/topics/db/managers/#from-queryset)
    ```python
     class CustomQuerySet(PolymorphicQueryset):
         ....
     class Model(PolymorphicModel):
         objects = PolymorphicManager.from_queryset(CustomQuerySet)
    ```
3. [CustomQuerySet.as_manager()](https://docs.djangoproject.com/en/5.2/ref/models/querysets/#as-manager)
    ```python
     class CustomQuerySet(PolymorphicQueryset):
         ....
     class Model(PolymorphicModel):
         objects = CustomQuerySet.as_manager()
    ```
4. Extend a custom manager class from the result of from_queryset:
    ```python
     class CustomQuerySet(PolymorphicQueryset):
         ....
     class CustomManager(PolymorphicManager.from_queryset(CustomQuerySet)):
         ....
     class Model(PolymorphicModel):
         objects = CustomManager()
    ```

**1** and **2** are functionally equivalent - ```as_manager()``` actually calls ```PolymorphicManager.from_queryset()``` under the covers. The only difference is that ```as_manager()``` will extend from ```PolymorphicManager``` by default unless you override the function in your ```QuerySet``` implementation. ```from_queryset()``` gives you the chance to combine any arbitrary subclass of ```PolymorphicManager``` with any arbitrary subclass of ```PolymorphicQueryset```.  You might use pattern **3** if you want to define functions on your manager that would not be available on a queryset returned by the manager (e.g. special versions of ```create()```).

Applying this to your code you could write:

```python
class MyBaseQuerySet(PolymorphicQuerySet):
    def filter_by_user(self, user):
        return self.all()


class MyBaseModel(PolymorphicModel):
    ...
    objects = MyBaseQuerySet.as_manager()


class MyChild1QuerySet(MyBaseQuerySet):
    def filter_by_user(self, user):
        return self.filter(fieldA__lt=user.profile.propA)


class MyChild1Model(MyBaseModel):
    fieldA = models.IntegerField()
    ...
    objects = MyChild1QuerySet.as_manager()


class MyChild2QuerySet(MyBaseQuerySet):
    def filter_by_user(self, user):
        return self.filter(fieldB__gt=user.profile.propB)


class MyChild2Model(MyBaseModel):
    fieldB = models.IntegerField()
    ...
    objects = MyChild2QuerySet.as_manager()
```

@dsummersl your example code does work. I tested it against the current release - it looks like a PR fixing this problem was added in 2013: #33

**Reply by fed239** [reply]
Hi @bckohan! Appreciate that you came back to this, the explanation makes sense. This discussion also brought back some good memories from my Django days, which were over like 5 years ago 🤠

**Reply by dsummersl** [reply]
thx @bckohan!
